emcache/autopipeline.py

```diff
@@ -0,0 +1,197 @@
+        except asyncio.TimeoutError as err:
+            # we need to raise this timeout to all of the desired waiters
+            for key in keys:
+                futures = self._ops.pop(key)
+                for future in futures:
+                    if not future.done():
+                        future.set_exception(err)
+
+            self._maybe_signal_termination()
+
+            # nothing else needs to be done since we finished with a timeout
+            # which invalidates the whole batch
+            return
+
+        returned_keys, values, flags, cas = results
+
+        # First we wake up waiters which we got a result
+        for idx in range(len(returned_keys)):
+            value = values[idx]
+            flags = flags[idx] if self._return_flags else None
+            cas = cas[idx] if self._return_cas else None
+
+            futures = self._ops.pop(returned_keys[idx])
+            for future in futures:
+                if not future.done():
+                    future.set_result(Item(value, flags, cas))
+
+        # Return a None for the missing keys
+        for key in set(keys) - set(returned_keys):
+            futures = self._ops.pop(key)
+            for future in futures:
+                if not future.done():
+                    future.set_result(None)
+
+        self._maybe_signal_termination()
+
+    def _maybe_signal_termination(self) -> None:
+        if len(self._ops) == 0:
+            self._on_finish(self)
+
+
+class AutoPipeline(BaseAutoPipeline):
+
+    _inflight_batches: List[asyncio.Task]
+    _pending_ops: Dict[bytes, List[asyncio.Future]]
+    _next_loop_handler: Optional[asyncio.Handle]
+    _client: Client
+    _cluster: Cluster
+    _loop: asyncio.AbstractEventLoop
+    _cas: bool
+    _return_flags: bool
+    _timeout: Optional[float]
+
+    def __init__(
+        self,
+        client: Client,
+        cluster: Cluster,
+        loop: asyncio.AbstractEventLoop,
+        *,
+        return_flags: bool,
+        return_cas: bool,
+        timeout: Optional[bool]
+    ) -> None:
+        self._pending_ops = {}
+        self._inflight_batches = []
+        self._next_loop_handler = None
+        self._loop = loop
+        self._client = client
+        self._cluster = cluster
+        self._return_flags = return_flags
+        self._return_cas = return_cas
+        self._timeout = timeout
+
+    def __del__(self):
+        if self._pending_ops:
+            logger.warning("Closing an autopipeline instance with pending operations to be batched")
+
+        if self._inflight_batches:
+            logger.warning("Closing an autopipeline instance with inflight batches")
+
+    def get(self, key: bytes) -> Awaitable[Item]:
+        if self._client.closed:
+            raise RuntimeError("Emcache client is closed")
+
+        if cyemcache.is_key_valid(key) is False:
+            raise ValueError("Key has invalid charcters")
+
+        future = self._loop.create_future()
```

Comment:
Should these have a timeout implementation? In case something goes wrong in the pending ops or if the queue isn't processed in time?

Reply:
hmm, I see the timeout is in the batch implementation. Should be ok.